In [1]:
import pandas as pd

dataset = {
  "x": [10, 20, 30, 30, 5],
  "y": [2, 50, 10, 15, 10]
}
df = pd.DataFrame(dataset)

print(df)

    x   y
0  10   2
1  20  50
2  30  10
3  30  15
4   5  10


In [2]:
data = []
for i in range(len(df)):
  data.append([int(df.loc[i, "x"]), int(df.loc[i, "y"])])

print(data)

[[10, 2], [20, 50], [30, 10], [30, 15], [5, 10]]


In [3]:
import math
import numpy as np

def euclidean_distance(a, b):
  return math.sqrt((a[0] - b[0])**2 + (a[1] - b[1])**2)

distance_matrix = []
for i in range(len(data)):
  temp = []
  for j in range(len(data)):
    dist = euclidean_distance(data[i],data[j])
    temp.append(dist)
  distance_matrix.append(temp)

distance_matrix = np.matrix(distance_matrix)
print(distance_matrix)

[[ 0.         49.03060269 21.54065923 23.85372088  9.43398113]
 [49.03060269  0.         41.23105626 36.40054945 42.72001873]
 [21.54065923 41.23105626  0.          5.         25.        ]
 [23.85372088 36.40054945  5.          0.         25.49509757]
 [ 9.43398113 42.72001873 25.         25.49509757  0.        ]]


**Interpretation**:

Most Similar: Obj3 and Obj4 are the most similar objects because they have the smallest Euclidean distance (5.00).

Least Similar: Obj1 and Obj2 are the most dissimilar because they have the largest distance (49.03)

In [6]:
import numpy as np

print("Similarity matrix: S = 1/(1 + D)")
similarity_matrix = 1/(1 + distance_matrix)
print(similarity_matrix)

Similarity matrix: S = 1/(1 + D)
[[1.         0.01998777 0.04436427 0.04023542 0.09584069]
 [0.01998777 1.         0.02367926 0.02673758 0.02287282]
 [0.04436427 0.02367926 1.         0.16666667 0.03846154]
 [0.04023542 0.02673758 0.16666667 1.         0.03774283]
 [0.09584069 0.02287282 0.03846154 0.03774283 1.        ]]


In [9]:
from math import *
import numpy as np

def cosine_similarity(x,y):
  magnitude_x = sqrt(x[0]**2 + x[1]**2)
  magnitude_y = sqrt(y[0]**2 + y[1]**2)
  if magnitude_x == 0 or magnitude_y == 0:
    return 0.0
  similarity = (x[0]*y[0] + x[1]*y[1]) / (magnitude_x * magnitude_y)
  return similarity

print("Cosine Similarity: ")
cosine_similarity_matrix = []
for i in range(len(data)):
  temp = []
  for j in range(len(data)):
    similarity = cosine_similarity(data[i],data[j])
    temp.append(similarity)
  cosine_similarity_matrix.append(temp)


cosine_similarity_matrix = np.matrix(cosine_similarity_matrix)
print(cosine_similarity_matrix)

Cosine Similarity: 
[[1.         0.54626778 0.99227788 0.96476382 0.61394061]
 [0.54626778 1.         0.64594224 0.74740932 0.99654576]
 [0.99227788 0.64594224 1.         0.98994949 0.70710678]
 [0.96476382 0.74740932 0.98994949 1.         0.8       ]
 [0.61394061 0.99654576 0.70710678 0.8        1.        ]]


In [16]:
import numpy as np

x = [data[i][0] for i in range(len(data))]
y = [data[i][1] for i in range(len(data))]

correlation = np.corrcoef(x, y)[0, 1]

print(f"Correlation coefficient between features x and y: {correlation:.4f}")

Correlation coefficient between features x and y: 0.1947
